client/src/js-sdk/sdk.js

```diff
@@ -1,49 +1,271 @@
+  // deep copy the card
+  var cardCopy = JSON.parse(JSON.stringify(card))
+  cardCopy["state.is_token"] = shouldCreateToken
+  cardCopy.id = uuidv4();
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [cardCopy.id]: cardCopy
+    })
+}
+
+export function flip(gameId, card) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  card["state.face_up"] = !card["state.face_up"]
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [card.id]: card
+    })
+}
+
+export function createToken(gameId, tokenScryfallId, targetPlayer, targetZone, power, toughness) {
+  var newCard = {
+    'id': uuidv4(),
+    'scryfall_id': tokenScryfallId,
+    'state.owner': targetPlayer,
+    'state.controller': targetPlayer,
+    'state.zone': targetZone,
+    'state.position': 0,
+    'state.tapped': false,
+    'state.face_up': true,
+    'state.clone_of': false,
+    'state.is_morph': false,
+    'state.is_token': true,
+    'state.power': power,
+    'state.toughness': toughness,
+    'attachments.counters': {},
+    'attachments.permanents': {}
+  }
+  db.doc(`Games/${gameId}/Players/${targetPlayer}/Zones/${targetZone}`)
+    .update({
+      [newCard.id]: newCard
+    })
+}
+
+export function setCardCounters(gameId, card, counterType, numCounters) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  var currentCounters = card["attachments.counters"]
+  currentCounters[counterType] = numCounters
+  card["attachments.counters"] = currentCounters
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [card.id]: card
+    })
+}
+
+export function setAttachedPermanents(gameId, card, attachedPermanents) {
+  var playerName = card["state.owner"];
+  var zoneName = card["state.zone"];
+  card["attachments.permanents"] = attachedPermanents
+  db.doc(`Games/${gameId}/Players/${playerName}/Zones/${zoneName}`)
+    .update({
+      [card.id]: card
+    })
+}
+
+//Player functions
+export function updateLife(gameId, player, newLife) {
+  db.doc(`Games/${gameId}/Players/${player}`)
     .update({
       life: newLife
     })
 }
 
+export function setPlayerCounters(gameId, player, counterType, numCounters) {
+  var currentCounters;
+  db.doc(`Games/${gameId}/Players/${player}`).get().then(function(doc) {
+    currentCounters = doc.data().counters
+    currentCounters[counterType] = numCounters
+    db.doc(`Games/${gameId}/Players/${player}`)
+      .update({
+        counters: currentCounters
+      })
+  });
+}
+
+//TODO we don't currently store any data for game status like this
+export function loseGame(player) {
+}
+export function winGame(player) {
+}
+export function drawGame() {
+}
+
+//Deck functions
+export function getAvailableDecks(user) {
+  var deckData;
+  db.doc(`Users/${user}`).get().then(function(doc) {
```

Comment:
My understanding is that this function as-is does not actually return any value. If you added a `return` to this line, it would return a `Promise` that resolves to the value returned by `Object.keys(deckData)`.